src/shogun/kernel/MaternKernel.cpp

```diff
@@ -0,0 +1,149 @@
+		result = (1 + ratio) * std::exp(-ratio);
+	}
+	else if (Math::fequals(nu, 2.5, std::numeric_limits<float64_t>::epsilon()))
+	{
+		float64_t ratio = (std::sqrt(5) * dist) / width;
+		result = (1 + ratio + ((5 * std::pow(dist, 2)) / (3 * std::pow(width, 2)))) * std::exp(-ratio);
+	}
+	// if none of the above was true we compute the value the expensive way
+	// in theory if nu is large we could calculate this with the squared exponential covariance function
+	else
+	{
+		float64_t ratio = std::sqrt(2 * nu) * dist / width;
+		// bessel function of the second kind
+#ifdef __clang__
+		// clang hasn't implemented mathematical special functions in the C++17 standard at this point
+		// note that yn casts nu to an integer
+		float64_t bessel = yn(nu, ratio);
+#else
+		float64_t bessel = std::cyl_neumann(nu, ratio);
+#endif
+		result = (std::pow(2, 1 - nu) / std::tgamma(nu)) * std::pow(ratio, nu) * bessel;
+	}
+
+	return result;
+}
+
+
+SGMatrix<float64_t> MaternKernel::get_parameter_gradient(Parameters::const_reference param, index_t index)
+{
+	require(lhs, "Left hand side features must be set!");
+	require(rhs, "Right hand side features must be set!");
+
+	if (param.first == "width")
+	{
+		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
+		std::function<float64_t(float64_t)> gradient_func;
+		
+		// the gradients Matern wrt width were computed with WolframAlpha
+		if (Math::fequals(nu, 0.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t width_squared = std::pow(width, 2);
+			gradient_func = [&](const float64_t& dist){ 
+				return (dist * std::exp(-dist / width)) / width_squared;
+			};
+		}
+		else if (Math::fequals(nu, 1.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t width_cubed = std::pow(width, 3);
+			gradient_func = [&](const float64_t& dist){ 
+					return (3 * std::pow(dist, 2) * std::exp(-(std::sqrt(3) * dist) / width)) / width_cubed; 
+			};
+		}
+		else if (Math::fequals(nu, 2.5, std::numeric_limits<float64_t>::epsilon()))
+		{
+			const float64_t width_power_4 = std::pow(width, 4);
+			gradient_func = [&](const float64_t& dist){ 
+				return (5 * std::pow(dist, 2) * std::exp(-(std::sqrt(5) * dist) / width) * (width + std::sqrt(5) * dist)) 
+						/ (3 * width_power_4);
+			};
+		}
+		else
+		{
+			gradient_func = [&](const float64_t& dist){
+				const auto ratio = (std::sqrt(2) * dist * std::sqrt(nu)) / width;
+#ifdef __clang__
+				auto bessel_nu_minus_1 = yn(nu - 1, ratio);
+				auto bessel_nu_plus_1 = yn(nu + 1, ratio);
+#else
+				auto bessel_nu_minus_1 = std::cyl_neumann(nu - 1, ratio);
+				auto bessel_nu_plus_1 = std::cyl_neumann(nu + 1, ratio);
+#endif
+				auto gradient = std::pow(2, (3 * nu) / 2 - 3 / 2);
+				gradient *= std::sqrt(nu) * dist;
+				gradient *= std::pow((std::sqrt(nu) * dist) / width, nu);
+				gradient *= bessel_nu_minus_1 - bessel_nu_plus_1;
+				gradient /= std::pow(width, 2) * std::tgamma(nu);
+
+				return gradient;
+			};
```

Comment:
@karlnapf as you can see the gradients are pretty complex... I used WolframAlpha to compute them, but would like to test them.. Is there a numeric gradient estimation in shogun I could use?

Reply:
I vaguely remember that we had something for that somewhere ....

Reply:
actually, why don't you implement the kernel in jax and generate some simple test cases in there (maybe in a gist that could be put as a comment in the unit test)

Reply:
I can check for matern 1/2, 3/2 and 5/2, but for the arbitrary order there is no way with JAX, as it hasn't implemented bessel of the second kind.. I'll have a look if I can test somewhere else.
Would be good to check the gradients though, like in sklearn unittests for kernels

Reply:
yeah checking for nontrivial cases already is useful.

Reply:
I think the finite diff stuff is somewhere in the nn code